### PR TITLE
[train][torchft] ray.train.report does not hang across replica group restarts

### PR DESCRIPTION
**Pull request by TimothySeah**
# Summary

Previously, Ray Train + torchft was incompatible with the `ray.train.report` API because if a worker died in the middle of a `ray.train.report` barrier, it would hang indefinitely.

There are two ways to solve this problem:
1) Clearing the `ray.train.report` state
2) Making the replacement worker rejoin the `ray.train.report` barrier.

The latter approach is tricky to get right. For example, look at how both workers are stuck in the following scenario:

```
for batch in batches:
	loss = ...
	loss.backward()
	optimizer.step()
	if manager.current_step() % 10:
		ray.train.report()
```

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-a8e24a3f-7fff-cf19-795b-873847307f5c"><div dir="ltr" style="margin-left:0pt;" align="left">
Worker 0 | Worker 1
-- | --
loss.backward() barrier | loss.backward() barrier
ray.train.report() barrier | dies
  | loss.backward() barrier

</div></b>

Therefore, I decided to go with the former approach. In fact, `torchft` itself employs a similar approach - when `loss.backward()` fails because a worker dies, it effectively skips over the batch without advancing the step count: https://github.com/meta-pytorch/torchft/blob/7eb7087885f7ffaf51148a9df3289d73b8709fbd/torchft/manager.py#L928. Furthermore, if we want to "replay" the report, we can do so using `manager.current_step`: https://github.com/ray-project/ray/pull/62651/changes#diff-935436cfeb39bc19b0652577dc71350a8352c3a8b797985ab30dab233930847dR173. Replaying the skipped batch is a separate problem; `torchft` recommends doing this with dataloader checkpointing, and Ray Data can potentially do this natively.

Note that `ray.train.report` has state in three places:
1) `SyncActor`: `replace_replica_group` now clear this with the new `reset` method.
2) `Worker`: `replace_replica_group` now clears the per-worker queues with the new `clear_result_queue` method
3) `ReportHandler`:  `replace_replica_group` now clears this with the `before_replica_group_shutdown`/`before_replica_group_start` callbacks.

Note that `test_report_with_worker_failure_before_sync` tests case 1 and `test_report_with_worker_failure_before_put_on_worker_queue` tests cases 2-3. 

# Testing

Unit tests + `test_torch_trainer` tests which are basically integration tests that use torchft.

### CODE REVIEW

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py [hunk 1]
```diff
@@ -116,3 +118,19 @@ def before_worker_group_shutdown(self, worker_group: WorkerGroup) -> None:
...
         self._num_workers = None
         self._training_report_queues = None
+
+    # --------------------------
+    # ReplicaGroupCallback
+    # --------------------------
+
+    def after_replica_group_start(self, replica_group: ReplicaGroup) -> None:
+        """Handle replica group start. Initialize internal states."""
+        self._num_workers += len(replica_group)
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/report_handler.py]: `after_worker_group_start` must always happen first. Clarified with assert.

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py [hunk 2]
```diff
@@ -116,3 +118,19 @@ def before_worker_group_shutdown(self, worker_group: WorkerGroup) -> None:
...
+        # TODO: it might be possible to reuse existing queues.
+        # For example, if 3/4 ddp workers reported a checkpoint, that checkpoint is usable.
+        self._training_report_queues = [deque() for _ in range(self._num_workers)]
+
+    def before_replica_group_shutdown(self, replica_group: ReplicaGroup) -> None:
+        """Handle replica group shutdown. Clear internal states."""
+        self._num_workers -= len(replica_group)
+        self._training_report_queues = [deque() for _ in range(self._num_workers)]
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/report_handler.py]: This cannot happen because the ray train controller is an asyncio loop. We can revisit if we make it multithreaded.

File: python/ray/train/v2/_internal/execution/context.py [hunk 3]
```diff
@@ -409,9 +410,18 @@ def report(
...
 
             # Sync the checkpoint dir name across ranks.
-            checkpoint_dir_name = self._sync_checkpoint_dir_name_across_ranks(
-                checkpoint_dir_name
-            )
+            try:
+                checkpoint_dir_name = self._sync_checkpoint_dir_name_across_ranks(
+                    checkpoint_dir_name
+                )
+            except ray.exceptions.RayTaskError:
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/context.py]: Done.

File: python/ray/train/v2/_internal/execution/worker_group/worker_group.py [hunk 4]
```diff
@@ -833,6 +833,19 @@ def _replace_replica_group_impl(self, replica_group_index: int):
                     w.distributed_context.world_rank, None
                 )
 
+        # Clear result queues on surviving workers to avoid mixed-generation
+        # results (some workers with old report, others with new report).
+        for i, rg in enumerate(self._replica_groups):
+            if i != replica_group_index and rg.is_active():
+                for w in rg.get_workers():
+                    ray.get(w.actor.clear_result_queue.remote())
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/worker_group/worker_group.py]: Done.

File: python/ray/train/v2/_internal/execution/worker_group/worker.py [hunk 5]
```diff
@@ -182,6 +182,23 @@ def poll_status(self) -> WorkerStatus:
...
 
+    def clear_result_queue(self) -> bool:
+        """Drain the result queue, discarding any pending training reports.
+
+        Returns:
+            True if the queue had at least one result, False if it was empty.
+        """
+        execution_context = get_train_context().execution_context
+        had_result = False
+        while True:
```
Comment (pseudo-rnd-thoughts) [inline python/ray/train/v2/_internal/execution/worker_group/worker.py:193]: Is it not possible to use a for loop for this or `while not execution_context.result_queue.empty()`. While true seems a tad excessive

File: python/ray/train/v2/tests/test_data_parallel_trainer.py [hunk 6]
```diff
@@ -283,6 +286,116 @@ def test_sigint_abort(spam_sigint):
...
+        scaling_config=ScalingConfig(num_workers=2),
+        run_config=RunConfig(
+            storage_path=str(tmp_path),
+            failure_config=FailureConfig(max_failures=1),
+        ),
+    )
+    result = trainer.fit()
+    assert result.error is None
+    assert result.metrics["iter"] == 2
+    assert len(result.best_checkpoints) == 1
```
Comment (pseudo-rnd-thoughts) [inline python/ray/train/v2/tests/test_data_parallel_trainer.py:325]: Do we only get one checkpoint at the end before report 1 isn't reported due to the worker failure?

File: python/ray/train/v2/tests/test_data_parallel_trainer.py [hunk 7]
```diff
@@ -283,6 +286,116 @@ def test_sigint_abort(spam_sigint):
...
+
+
```
Comment (pseudo-rnd-thoughts) [inline python/ray/train/v2/tests/test_data_parallel_trainer.py:398]: How does this interact with validation functions? As they are async then torchft doesn't interact with it?

**Review by pseudo-rnd-thoughts — APPROVED** [review]
Overall looks good, I just have a couple of questions

File: python/ray/train/v2/_internal/execution/worker_group/worker.py (hunk 5 above)
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/worker_group/worker.py:193]: I agree it looks weird, but apparently `Queue.empty` isn't reliable in a concurrent context: https://docs.python.org/3/library/queue.html#queue.Queue.empty

File: python/ray/train/v2/tests/test_data_parallel_trainer.py (hunk 6 above)
Comment (TimothySeah) [reply python/ray/train/v2/tests/test_data_parallel_trainer.py:325]: Yeah. Report 1 was stuck in the barrier, the worker died, the controller cleared out the barrier, and then we hit the report 2 barrier.

File: python/ray/train/v2/tests/test_data_parallel_trainer.py (hunk 7 above)
Comment (TimothySeah) [reply python/ray/train/v2/tests/test_data_parallel_trainer.py:398]: Validation basically has 2 steps: 1) Workers form a report barrier to "commit" the checkpoint to the controller 2) Controller manages validation for the committed checkpoint. Torchft affects the first part, which is why this PR is clearing out the barrier. It doesn't affect the second part because the controller manages the validation task independently of the worker group.

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py [hunk 8]
```diff
@@ -116,3 +121,21 @@ def before_worker_group_shutdown(self, worker_group: WorkerGroup) -> None:
...
+        self._num_workers += len(replica_group)
+        # TODO: it might be possible to reuse existing queues.
+        # For example, if 3/4 ddp workers reported a checkpoint, that checkpoint is usable.
+        self._training_report_queues = [deque() for _ in range(self._num_workers)]
+
+    def before_replica_group_shutdown(self, replica_group: ReplicaGroup) -> None:
+        """Handle replica group shutdown. Clear internal states."""
+        self._assert_initialized()
+        self._num_workers -= len(replica_group)
+        self._training_report_queues = [deque() for _ in range(self._num_workers)]
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/report_handler.py]: Done - counted from worker group instead.

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py [hunk 9]
```diff
@@ -10,25 +11,28 @@
...
     def __init__(self, report_callbacks: List[ReportCallback]):
-        # Number of workers in the current worker group. It is initialized
-        # to be None. It is set to the number of workers when it receives the
-        # worker group status for the first time.
-        # When a worker group shutdown, self._num_workers is set to None,
-        # waiting to be updated when a new worker group status is received again.
-        self._num_workers: Optional[int] = None
+        # We set the worker group after it has been started and remove it after it
+        # has been shut down.
+        self._worker_group: Optional[WorkerGroup] = None
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/execution/checkpoint/report_handler.py:25]: why do we need to change from tracking num_workers to the actual worker group? we don't use the worker group at all and only use len(worker_group) anyways

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py [hunk 10]
```diff
@@ -106,13 +106,24 @@ def after_worker_group_poll_status(
...
+    # --------------------------
+    # ReplicaGroupCallback
+    # --------------------------
+
+    def after_replica_group_start(self, replica_group: ReplicaGroup) -> None:
+        """Handle replica group start. Initialize internal states."""
+        self._assert_initialized()
+        # TODO: it might be possible to reuse existing queues.
+        # For example, if 3/4 ddp workers reported a checkpoint, that checkpoint is usable.
+        self._training_report_queues = [deque() for _ in range(len(self._worker_group))]
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/execution/checkpoint/report_handler.py:129]: Does this handle the case where one of the replica groups goes down after the barrier but before the report? In which case the controller is left with a partial report where some of these report queues aren't empty?

File: python/ray/train/v2/tests/test_data_parallel_trainer.py [hunk 11]
```diff
@@ -283,6 +286,116 @@ def test_sigint_abort(spam_sigint):
     process.join()
 
 
+def test_report_with_worker_failure_before_sync(tmp_path):
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_data_parallel_trainer.py]: can we put these replica group tests in a separate file?

**Review by justinvyu — APPROVED** [review]
Looks good overall!

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py (hunk 9 above)
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/report_handler.py:25]: This isn't needed for the "fixed worker group" case that we are currently optimizing for, but if we want to support torchft-compatible elastic training, we need a way to update `num_workers` when replica groups join/leave while avoiding incorrect state updates from retries: https://github.com/ray-project/ray/pull/62651#discussion_r3090554008

File: python/ray/train/v2/_internal/execution/checkpoint/report_handler.py (hunk 10 above)
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/report_handler.py:129]: Yeah basically. There are three places where `ray.train.report` could leave partial state:
1) `SyncActor` barrier
2) On the `worker` but not pulled to the controller
3) On the `controller`

This handles case 3.

File: python/ray/train/v2/tests/test_data_parallel_trainer.py (hunk 11 above)
Comment (TimothySeah) [reply python/ray/train/v2/tests/test_data_parallel_trainer.py]: Done.

File: python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py [hunk 12]
```diff
@@ -80,6 +88,8 @@ def _setup_or_validate_collective_op(self, world_size: int):
         if self._world_size == 0:
             self._world_size = world_size
             self._sync_start_times = [None] * world_size
+            # Clear any stale reset flag from a previous barrier cycle.
+            self._reset = False
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py]: Added TODO for now since I think this is unlikely to happen in practice. Basically we need:
1) worker dies
2) replace_replica_group calls reset
3) some but not all workers leave the context manager which decrements the counter (counter is not 0)
4) worker calls new broadcast. Undefined behavior because state is not cleared.

This is unlikely to happen because:
1) It requires 2 `ray.train.report`s in immediate succession. But why do this when you can just combine into 1 report?
2) In practice I've noticed that when you notify asyncio coroutines and then add a new asyncio call, the notified coroutines generally finishb efore the new asyncio call starts. 

It's tricky to fix because we essentially want the contextmanager to wait for the counter to be 0, but then the first function that passes that barrier will make the counter nonzero again.

File: python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py [hunk 13]
```diff
@@ -185,6 +210,11 @@ async def broadcast_from_rank_zero(
                         warn_interval_s=self._warn_interval_s,
                         timeout_s=self._timeout_s,
                     )
+                    if self._reset:
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py:212]: I think this is fine - that data you return is from a successful barrier.

File: python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py]: Wait actually this was easy to fix. We just need to wait for reset to be False before starting.

File: python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py [hunk 14]
```diff
@@ -71,12 +78,15 @@ def _clear_states(self):
         if self._counter == 0:
             self._reduced_data = None
             self._world_size = 0
+            self._reset = False
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/_internal/execution/checkpoint/sync_actor.py:81]: Whoops - done.
